### ISSUE TITLE
Tab Sync Across Devices

### ISSUE THREAD

**Issue by seniorm0ment** [mbnuqw/sidebery#1372]
## Description
I use Sideberry on all my devices, and I have Firefox Sync enabled for Tabs.
It would be very nice to be able to have it Sync across devices.

Maybe this would be best done by making it a specific "Tabs panel" for synced tabs. Like when you create a new tabs panel, just offer a checkbox to "sync across devices".  Then users can create multiple tab panels (like I have a panel for work, and personal) that all sync across devices. Then they can not check the box to have device specific tab panel.

When Firefox sync syncs the tabs, it would show up on the other device.
So I can go from my desktop to laptop seamlessly and have all my tabs there.

I also use Mull on Android, obviously Sideberry doesn't apply, but maybe if it could offer some way to view those tabs with ease too. Maybe an option when creating a tab panel to "show android tabs as a folder in this panel" so they are in a folder and you just open it and it is still synced across devices or something. Not too sure.

**Comment by hobgoblina** [comment]
This would be great. Even just syncing snapshots would be solid. 

Additionally, it'd be great if we could define a local directory to store synced data in. This would remove data size constraints and enable us to use methods like Syncthing to sync across devices. I don't use firefox's sync feature because I don't want to send me browsing data to a third party, so this is a necessary feature for me to have sync functionality.

**Comment by seniorm0ment** [comment]
You can actually selfhost Firefox Sync btw

But I would not be against an additional option to define the local directory either (this would actually be nice for Snapshots imo)

However, I feel Syncthing gets buggy with more realtime stuff like this, I use it for Keepass and it doesn't always sync properly,  often conflicts, etc. Overall its pretty good, but its the times it doesn't that's annoying. 
I'd definitely like it to use Firefox Sync, however again I'm not against an _option_  to define a directory either.

**Comment by alwintang** [comment]
The ability to specify a local directory for snapshots seems to be available now. You can simply configure it to automatically export to a JSON file when generating a snapshot. The exported file is stored in the browser's download directory, and the directory containing the JSON file can be synchronized through the cloud disk. Other computers can import it in the snapshot panel with just a single manual step.

On another note, is there a one-step button to access the snapshot panel? Currently, it appears that only a shortcut key method is available. It would be more convenient for me to have a small **_button_** added.

**Comment by hobgoblina** [comment]
"stored in the download directory" isn't really "the ability to select a local folder" ... would be nice if we could select a directory other than `Downloads` or a subdirectory of it

**Comment by alwintang** [comment]
Yes,this allows us to limitedly choose a local folder. My sync stream works simplely, I think it is acceptable.

**Comment by hobgoblina** [comment]
right, I just don't want it in my downloads folder for a variety of reasons. acceptable for you doesn't mean it is for everyone, hence the request.

**Comment by mertyn** [comment]
something like this would be a really great feature, i can relate with the multiple devices!
for me _personally_, just syncing using firefox sync and either doing snapshots or downright syncing the whole state would be ideal. having the option to sync using snapshot files aswell would be great too. with this however, auto import and the option of auto loading the most recent snapshot would make it more frictionless.
i don't know if giving all of these options at the same time is viable to be implemented, but given the customizable nature of the whole extension, it would be really nice to see all available.

**Comment by seniorm0ment** [comment]
Yes not to bump my own thread, but the more I am using my laptop for business lately, the more I find myself wishing I could just come home and shut the lid on my laptop, and hop on my desktop and not have to open up a million of the tabs I had open already on my laptop. It would be so nice if they were just there already synced and ready to go. Then when I don't want to be on my desk, I can just grab my laptop and jump on the couch and be where I left off from my desktop.

Maybe an option to right click > "open/close tab locally" might be something to consider alongside this.

But yeah just a basic sync would be so nice.

**Comment by artfox3** [comment]
I guess the easiest way to do this for now is using the bookmarks, at the end of every session in one device, i save the panel to a folder bookmarks, and because bookmarks are synced, on the other device i just open the panel from that folder.
But it isn't intuitive, as you have to save manually and open manually every time, what would be great is to have an option to save automatically to bookmarks, the same we have for snapshots, and another option to update automatically a panel from its bookmark folder for example.

**Comment by daffydock** [comment]
If I may be able to add, I had recently been using Tab Stash and its implementation to sync tabs across devices is pretty good. 

https://addons.mozilla.org/en-US/firefox/addon/tab-stash/

It saves all tab groups as bookmark folders under a specific main folder: 'Tab Stash.' So, an install new or otherwise on any browser will automatically look into that folder for any existing tab groups and sets them/syncs them up, as per your syncing cycle or tool. 

It uses FF Sync but essentially any sync service/app will work. I, like others, do not use FF's implementation but Nextcloud Bookmarks and its extension to sync bookmarks across devices through my instance and it works great. 

However, I still use SB as well, since the UI, options and general usability of SideBerry is superb in comparison. No dig at TD, since TD feels far more like an archiving extension over the more all-rounder that SB is. All you have to do is save them, which is something we all can already do in SB. Syncing the JSON file is also a possible but it is really not efficient in comparison to syncing tab folders. 

It might be a tad limiting to have all tab groups under one a specific folder for some, yet if the end goal is to have a portable experience, whether ones uses FF sync or another syncing service, to me it seems quite the worthwhile trade off.

I am not sure if this can be implemented in that manner. For now, I save my tab groups as specific folders and the bookmarks get synced but I have not yet tried to open them on a different device/SB install, as I do not want to risk overwriting them if I create a new group on different install without FF Sync, of if they will repopulate if I name the Tab Group the same as the existing Bookmark folder. 

Either way, thank you for a great extension. 

Update: Out of curiosity, tried a while ago the DIY approach mentioned above after originally writing this, of syncing SB via bookmarks only and well, it messed up my bookmarks. Fixed it but would not recommend, until SB incorporates a more TS approach that is built-in or just use the original way of export/import.

**Comment by lufixSch** [comment]
I really like the tab-stash solution because it also allows me to go through my tabs/groups on mobile (ios) even though there is no support for extensions. Would really love to see that added to Sidebery.

**Comment by hollisticated-horse** [comment]
I was about to post : 

> Out of curiosity, is there something technical halting implementation of this feature ? Or is it to do with project and time management ? 

But seems that we have an answer from @mbnuqw [here](https://github.com/mbnuqw/sidebery/discussions/1488)

> [#54](https://github.com/mbnuqw/sidebery/issues/54), It's planned but will probably require some third-party (or self-hosted) service because of Firefox [limitations](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/sync#storage_quotas_for_sync_data). Not sure about when it will be implemented, since I don't need that feature myself.

They also mention some workarounds.

Maybe we continue the conversation there ? 

Anyhow, +1 for Sync support (like Tabstash or other), and thank you soo much for that extension. 

Would totally donate if a non-crypto system were to be provided :grin:

**Comment by davidjhi** [comment]
+1 for sync support of tabs
+1 of additional system of donations!

**Comment by mertyn** [comment]
i actually did not realize that one can save and restore tab panels to and from bookmarks. that feature is pretty nice, but not quite what i (and probably others in need of sync) want. i also had a look at tab stash, but that also does not achieve what i would want, because it bunches together _all_ tabs present.
if the system for using bookmarks is already in place, then it should be possible to use that for saving and restoring _all_ tab panels. and also not save the panels as new folders every time, but overwrite the existing one. making that process automatic or semi-automatic would then be really convenient

**Comment by Terkyz** [comment]
+1 asking for this feature.

**Comment by migmanu** [comment]
Yeah +1 here too!

**Comment by yves-amevoin** [comment]
+1 here too!

**Comment by c0d3h4x0r** [comment]
This should be considered more important than any other feature request or non-breaking bug. This is a basic, fundamental feature that every user naturally expects to be present. It happens to be the one missing feature that is an absolute dealbreaker for me trying out this extension.

**Comment by TheFriendlyGoat** [comment]
I would switch to Firefox if auto tab syncing was implemented. I currently use the Workspaces feature of Microsoft Edge that syncs tabs across devices. It is so useful when switching from my laptop to my desktop, I can't use a browser without it.

I do love Sideberry, but can't use it without sync. Please look at implementing this ASAP.

**Comment by davidjhi** [comment]
I'm very happy using Sidebery.

I would love to have this functionality, but as the developer mentions, it's not as straight forward as it can be. Also, i think Sidebery its just  a single person project, that does this in his free time.

So I would encourage people to donate something so he can consider this, as something that people really need. I just sent around 26usd, not that much, but It's something:

![imagen](https://github.com/user-attachments/assets/e2f59f27-ae5f-48b7-83ed-5835b0f78966)

I think the simpler method would be a local backup, so people can use sync tools to mantain the device's tabs updated, for example Syncthing /  Nextcloud.

PD: 
There's been like 4 - 5 times that Firefox gets updated by snap (Ubuntu), and somehow it gets disconnected from my Firefox User. When i log-in again, the device is recognized as something different, so all my tabs get lost. That its super annoying. Having at least a backup of that would be great.

**Comment by mbnuqw** [comment]
@davidjhi, Thank you for donation, I'll rise the priority of this task. Although, I need a few days to catch up with the latest activity on this project, merge some PRs, fix some nasty bugs and probably make the new AMO release with the latest commits. After that I'll check what I can do here. Probably will test/[implement] multiple options: Google Drive (and other) API, #1076, self-hosted sync server, etc...

**Comment by brandonpollack23** [comment]
+1 will donate some for this once my account is verified (haha dont have crypto accounts)

**Comment by TheFriendlyGoat** [comment]
Just donated a little towards this feature. I understand if it's complicated or takes a bit, but would love to see it. Great plugin too!

![Screenshot_20240821_002401_Edge.jpg](https://github.com/user-attachments/assets/7f3997c2-7877-4b59-b1a0-1b64e9eb0f22)

**Comment by rsenna** [comment]
+ 100 from my side!

Btw, probably a very obvious reminder, but I guess still worth mentioning: I believe there are many ways a "simple" synchronization could go wrong. For example, ff two or more users are sharing the same Firefox account and try to make different changes to tabs panels, folders or tabs at the same time. That would very probably cause synchronization conflicts, which would have to be handled adequatedly.

Otherwise, a naive implementation could easily provoke different problems, such as losing tabs ordering, losing whole tabs, or issues with the folder structure.

**Comment by daffydock** [comment]
Yes, there are many ways but was wondering, can SB save tabs only on panels? As in in its own database, if that is the case? Can there not be an option to save tabs as bookmarks in a folder? Say, named, 'Sideberry Tabs' with subfolders, instead of panels, etc. Let FF sync do the job of the sync between devices. So, SB does not have to do the job. 

As far as as numerous people using the same FF account, how many people are actually in that boat? Percentage would be small, although a type of priority set could be used. Or a warning to the end user to literally state the limitations of the sync. No extension can cover every singular use case. 

Syncing among devices can be tricky as it is, adding multiple users using the same Profile at the same time can create issues in many extentions, not just SB. I get the impression that is not the main use case for FF profiles. Changing tabs at will, in FF profiles usually meant to be used by one user at the time, should not be the reason why a sync solution should be stopped from being implemented. Many sync apps do seem to have the possible issue with asynchronous syncs, so leave it to FF Sync to deal with.  No reason tor reinvent the wheel on  that front, since they do a good job in syncing bookmarks and open tabs, already. 

Nextcloud Bookmarks implements a very simple set of settings as to set a hiarchy/options as to how bookmarks are to be synched and when. If it is okay with you, you may look at how they do it for possible ideas. 

Either way, thank you for such a great extension, and here is to everyone having a great 2025! Cheers.

**Comment by mbnuqw** [comment]
Update (a bit late) on this issue.

With the release of v5.3.1 on addons.mozilla.org you can test how experimental synchronization of tabs works. It works via Google Drive, so to activate it, open Sidebery Settings / Sync and turn on the "Use Google Drive" setting.

A new panel for viewing and working with synchronized data has been added, which can be placed in the sidebar (check the Settings / Navigation bar) or opened via a shortcut.

To send selected tabs to the sync storage (Google Drive) you can use the context menu option (you need to activate it in the Settings / Menu Editor) or drag-n-drop the tabs to the sync panel icon (in the sidebar).

https://github.com/user-attachments/assets/db41079a-1acc-4b45-8873-0b51f60c6dcc

Plans regarding further development of this function:
- It is planned to add more sync providers (in addition to current Firefox Sync and Google Drive), including self-hosted solutions.
- Automatic saving to sync storage
  - Per panels
  - Per window
  - All tabs
- Synchronize last N snapshots
- Improvements to UX when working with the sync panel
- Data encryption

Comments and feedback are welcome.

**Comment by megamorphg** [comment]
I'd really like to test this feature and move it forward. But I'm stuck at the very start because Google Drive keeps asking for authentication repeatedly.

Having Sidebery actually sync tabs across two devices in the specific named windows I select (the window titles are prefixed, but any other way of mapping window is acceptable), including panels, containers, nesting, over-ridden window titles, group page, panels that are tied to bookmarks, etc
Having some windows automatically update their windows and some windows "offer" to update the windows (or not offer but allow user to on-demand decide) are all important features, too.

**Then I could simply work on my desktop and then jump over to my laptop and have the same tabs.**
